### PR TITLE
Add some text regarding using passwords with Fernet.

### LINKED ISSUE #2160: Discuss KDF usage with Fernet
Based on #2159, it would probably make sense to mention using HKDF in the docs with Fernet. That way users who are trying to encrypt without using a key stored on disk somewhere have clear guidance.

### PR DESCRIPTION
**Pull request by Ayrx**
Fixes #2160.

### CODE REVIEW

File: docs/fernet.rst [hunk 1]
```diff
@@ -106,6 +106,43 @@ has support for implementing key rotation via :class:`MultiFernet`.
...
+    )
+    key = base64.urlsafe_b64encode(kdf.derive(password))
+    f = Fernet(key)
+
+In this scheme, the salt has to be stored in a retrievable location in order
+to derive the same key from the password in the future.
+
+The iteration count used should be adjusted to be as high as your server can
+tolerate. A good default is at least 100k iterations which is what Django
+`recommends`_.
```
Comment (reaperhulk) [inline docs/fernet.rst]: We should note the year this recommendation was made (2014) so it becomes obvious in the future when this needs updating.

File: docs/fernet.rst [hunk 2]
```diff
@@ -106,6 +106,43 @@ has support for implementing key rotation via :class:`MultiFernet`.
 
     See :meth:`Fernet.decrypt` for more information.
 
+
+Using passwords with Fernet
+---------------------------
+
+It is possible to use passwords with Fernet. To do this, you need to run the
+password through a key derivation function like
```
Comment (alex) [inline docs/fernet.rst]: `s/link/such as/`

File: docs/fernet.rst [hunk 3]
```diff
@@ -106,6 +106,43 @@ has support for implementing key rotation via :class:`MultiFernet`.
...
+:class:`~cryptography.hazmat.primitives.kdf.pbkdf2.PBKDF2HMAC`:
+
+.. code-block:: python
+
+    import base64
+    import os
```
Comment (alex) [inline docs/fernet.rst]: Newline between import sections.

File: docs/fernet.rst [hunk 4]
```diff
@@ -106,6 +106,43 @@ has support for implementing key rotation via :class:`MultiFernet`.
...
+    from cryptography.hazmat.primitives import hashes
+    from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
+    from cryptography.hazmat.backends import default_backend
```
Comment (alex) [inline docs/fernet.rst]: This import shoudl come before the primitives ones

File: docs/fernet.rst [hunk 5]
```diff
@@ -106,6 +106,43 @@ has support for implementing key rotation via :class:`MultiFernet`.
...
+    from cryptography.fernet import Fernet
```
Comment (alex) [inline docs/fernet.rst]: f comes before h :-)

File: docs/fernet.rst [hunk 6]
```diff
@@ -106,6 +106,43 @@ has support for implementing key rotation via :class:`MultiFernet`.
...
+        backend=default_backend
+    )
+    key = base64.urlsafe_b64encode(kdf.derive(password))
+    f = Fernet(key)
+
+In this scheme, the salt has to be stored in a retrievable location in order
+to derive the same key from the password in the future.
+
+The iteration count used should be adjusted to be as high as your server can
+tolerate. A good default is at least 100k iterations which is what Django
```
Comment (alex) [inline docs/fernet.rst]: Please write out 100k as "100,000"

**Comment by alex** [comment]
Should this section mention other KDFs such as bcrypt?

**Comment by Ayrx** [comment]
@alex I don't think so, at least until cryptography itself supports them.

**Comment by alex** [comment]
pyca maintains `bcrypt` though.

On Sat, Aug 29, 2015 at 10:30 AM, Terry Chia <notifications@github.com>
wrote:

> @alex <https://github.com/alex> I don't think so, at least until
> cryptography itself supports them.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/2304#issuecomment-135991669>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

**Comment by reaperhulk** [comment]
Looks like you need to add `bcrypt` to the spelling wordlist.

**Pull request merged**